### PR TITLE
NIAD-3067: codeable concept changes

### PR DESCRIPTION
**Pull request by ole4ryb**
## What

Codeable concept comparison

## Why

Please include details of the reasoning for these changes

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Internal change (non-breaking change with no effect on the functionality affecting end users)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have updated the [Changelog](/CHANGELOG.md) with details of my change in the UNRELEASED section if this change will affect end users
- [ ] A corresponding change has been made to the [Mapping Documentation repository][mapping-docs]

[mapping-docs]: https://github.com/NHSDigital/patient-switching-adaptors-mapping-documentation

### CODE REVIEW

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        cr2.setCode("QualifierCode");
+        cr2.setCodeSystem("QualifierCodeSystemCode");
+        cr2.setDisplayName("QualifierDisplayName");
+        c2.getQualifier().add(cr2);
+
+        assertTrue(CodeableConceptUtil.compareCodeableConcepts(c1, c2));
+    }
+
+    @Test
+    public void compareCodeableConceptsWithTwoTransactionsPairsNotIdenticalTranslations() {
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java:90]: This test name is a mouthful, and also has the word transaction in it.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        translationCD1.setCode("161586000");
+        translationCD1.setCodeSystem("2.16.840.1.113883.2.1.3.2.4.15");
+        translationCD1.setDisplayName("H/O: injury");
+        c1.getTranslation().add(translationCD1);
+
+        CD translationCD2 = new CD();
+        translationCD2.setCode("161586007");
+        translationCD2.setCodeSystem("2.16.840.1.113883.2.1.3.2.4.17");
+        translationCD2.setDisplayName("H/O: injury7");
+        c1.getTranslation().add(translationCD2);
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java:108]: Did you intentionally not add the translations to the Codable Concept. If so why create them in the first place?

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    @Test
+    public void compareCodeableConceptsWithSecondNullParam() {
+        CD c1 = new CD();
+        assertFalse(CodeableConceptUtil.compareCodeableConcepts(c1, null));
+    }
+
+    @Test
+    public void compareCodeableConceptsWithOneTransaction() {
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java:30]: Transaction typo.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        translationCD4.setDisplayName("H/O: injury7");
+
+        c2.getTranslation().add(translationCD3);
+        c2.getTranslation().add(translationCD4);
+
+        assertFalse(CodeableConceptUtil.compareCodeableConcepts(c1, c2));
+    }
+
+    @Test
+    public void compareCodeableConceptsWithUniqualNumberOfTransactionPairs() {
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java:133]: Unequal typo.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        translationCD1.setCode("161586000");
+        translationCD1.setCodeSystem("2.16.840.1.113883.2.1.3.2.4.15");
+        translationCD1.setDisplayName("H/O: injury");
+        c1.getTranslation().add(translationCD1);
+
+        CD translationCD2 = new CD();
+        translationCD2.setCode("161586007");
+        translationCD2.setCodeSystem("2.16.840.1.113883.2.1.3.2.4.17");
+        translationCD2.setDisplayName("H/O: injury7");
+        c1.getTranslation().add(translationCD2);
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java:151]: Again why create these translations? I don't think they're being used?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtil.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+                && Objects.equals(c1.getOriginalText(), c2.getOriginalText())
+                && compareCodeSubElements(c1.getQualifier(), c2.getQualifier())
+                && compareCodeSubElements(c1.getTranslation(), c2.getTranslation());
+    }
+
+    private static Boolean compareCodeSubElements(List<? extends CD> c1, List<? extends CD> c2) {
+
+        if (CollectionUtils.isEmpty(c1) || CollectionUtils.isEmpty(c2) || c1.size() != c2.size()) {
+            return CollectionUtils.isEmpty(c1) && CollectionUtils.isEmpty(c2);
+        }
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtil.java:31]: I feel like our tests will continue to pass, and the code simplier, if we just replaced this with:

```suggestion
        if (c1.size() != c2.size()) {
            return false;
        }
```

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java (hunk 1 above)
Comment (ole4ryb) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java]: fixed

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java (hunk 3 above)
Comment (ole4ryb) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java]: fixed

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java (hunk 4 above)
Comment (ole4ryb) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java]: fixed

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtil.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+        for (CD c1Element : c1) {
+            boolean matchFound = c2
+                    .stream()
+                    .anyMatch(c2Element -> compareCodeableConcepts(c1Element, c2Element));
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtil.java:36]: Do we have any tests which cover the ordering of two sets of Translations being different?

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java (hunk 2 above)
Comment (ole4ryb) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java:108]: Translation expects to receive items of type CD which is different type to CodeableConcept. Symantically they are similar  in the way how we use the fields(Code, CodeSystem and DisplayName fields) but they are still different objects with different names -> CD vs. CodeableConcept

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java (hunk 5 above)
Comment (ole4ryb) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtilTest.java:151]: they are used

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtil.java (hunk 6 above)
Comment (ole4ryb) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/CodeableConceptUtil.java]: I've done the change that was suggested

**Review by adrianclay — APPROVED** [review]

**Pull request merged**